department-of-veterans-affairs/vets-website
# [Bug] Fix U.S. address UX in form 686C-674

## Original Ticket

https://github.com/department-of-veterans-affairs/va.gov-team/issues/42216

## URL of change

- `/view-change-dependents/add-remove-form-21-686c/veteran-address`
- `/view-change-dependents/add-remove-form-21-686c/current-marriage-address`
- `/view-change-dependents/add-remove-form-21-686c/add-child/0/additional-information`
- `/view-change-dependents/add-remove-form-21-686c/report-674-student-address`
- `/view-change-dependents/add-remove-form-21-686c/686-stepchild-no-longer-part-of-household/0`

## Background

[A Report from the Pittsburgh Regional Office](https://dsva.slack.com/archives/C5AGLBNRK/p1654182876865439) stated:
> Veterans/claimants that reside within the United States are having instances of an APO, DPO, 
or FPO erroneously assigned at the time they submit a claim. Pittsburgh Regional Office conducted a fairly thorough investigation of the issue on their side of the integration: https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/assets/False%20Foreign%20White%20Paper_V3.docx

## Steps to test

To test an address:

1. Enter any U.S. city & state
2. Select the "I live on a United States military base..." checkbox
3. Enter any military city & state address (e.g. APO + AA)
4. Uncheck the "I live on a United States military base..." checkbox
5. The city & state from step 1 should be restored.
6. Repeatedly checking the checkbox will clear the military city & state (and show an error) or restore the non-military city & state

To access & test the 5 pages:

1. Start a review instance or pull in the branch locally
2. Go to `http://localhost:3001/view-change-dependents/add-remove-form-21-686c/introduction`
3. Log in to test user 228 (Mark Webb)

### Add Spouse
1. On the `86-options-selection` page with 8 checkboxes, choose "Add a spouse"
2. Navigate to the `veteran-address` page and test the address (if needed)
3. Fill out spouse info, marriage info page and navigate to `current-marriage-address` page
4. Choose "No" for "Does your spouse live with you?" and then test the address

### Add Child
1. On the `86-options-selection` page with 8 checkboxes, choose "Add a child or children under 18..."
2. Navigate to the `veteran-address` page and test the address (if needed)
3. Fill out fields & navigate to `add-child/0/additional-information` page
4. Choose "No" for "Does this child live with you?" and then test the address

### School
1. On the `86-options-selection` page with 8 checkboxes, choose "Add a child 18... who’ll be attending school..."
2. Navigate to the `veteran-address` page and test the address (if needed)
3. Fill out fields & navigate to `report-674-student-address` page
4. Test the student address

### Remove a stepchild
1. On the `86-options-selection` page with 8 checkboxes, choose "Remove a stepchild..."
2. Navigate to the `veteran-address` page and test the address (if needed)
3. Fill out fields & navigate to `686-stepchild-no-longer-part-of-household/0` page
4. Test the student address

## Code changes

- Updated form 686c-674 address schema to use available constants
- Added logic for maintaining non-military city & state values in a common `updateFormData` function
- Added `updateFormData` to the five address fields that include the military address checkbox
- Added page index to platform `FormPage` & `ReviewCollapsibleChapter` components
- Updated testing tool utility (`schemaform-utils`) to call the `updateFormData` function when provided (thanks @jerekshoe!)
- Added platform unit tests

## How was your code tested

- Both manually and with unit tests

## Acceptance criteria
- [x] Update platform FormPage to include index in `updateFormData`
- [x] Update all addresses in 686c & 674, that include military checkbox 

## Definition of done
